File: runtime/net/go/net_runtime.go
package gonet

import (
	"net"
	"sync"
	"time"
)

// Connection and listener handle management
var (
	connections      = make(map[int]net.Conn)
	listeners        = make(map[int]net.Listener)
	nextConnHandle   = 1
	nextListenHandle = 1
	connMutex        sync.Mutex
	listenMutex      sync.Mutex
)

// Dial connects to the address on the named network.
func Dial(network string, address string) (int, string) {
	conn, err := net.Dial(network, address)
	if err != nil {
		return -1, err.Error()
	}

	connMutex.Lock()
	handle := nextConnHandle
	nextConnHandle++
	connections[handle] = conn
	connMutex.Unlock()

	return handle, ""
}

// Listen announces on the local network address.
func Listen(network string, address string) (int, string) {
	listener, err := net.Listen(network, address)
	if err != nil {
		return -1, err.Error()
	}

	listenMutex.Lock()
	handle := nextListenHandle
	nextListenHandle++
	listeners[handle] = listener
	listenMutex.Unlock()

	return handle, ""
}

// Accept waits for and returns the next connection to the listener.
func Accept(listener int) (int, string) {
	listenMutex.Lock()
	l, ok := listeners[listener]
	listenMutex.Unlock()

	if !ok {
		return -1, "invalid listener handle"
	}

	conn, err := l.Accept()
	if err != nil {
		return -1, err.Error()
	}

	connMutex.Lock()
	handle := nextConnHandle
	nextConnHandle++
	connections[handle] = conn
	connMutex.Unlock()

	return handle, ""
}

// Read reads up to size bytes from the connection.
func Read(conn int, size int) ([]byte, int, string) {
	connMutex.Lock()
	c, ok := connections[conn]
	connMutex.Unlock()

	if !ok {
		return nil, 0, "invalid connection handle"
	}

	buf := make([]byte, size)
	n, err := c.Read(buf)
	if err != nil {
		return nil, 0, err.Error()
	}

	return buf[:n], n, ""
}

// Write writes data to the connection.
func Write(conn int, data []byte) (int, string) {
	connMutex.Lock()
	c, ok := connections[conn]
	connMutex.Unlock()

	if !ok {
		return 0, "invalid connection handle"
	}

	n, err := c.Write(data)
	if err != nil {
		return n, err.Error()
	}

	return n, ""
}

// Close closes the connection.
func Close(conn int) string {
	connMutex.Lock()
	c, ok := connections[conn]
	if !ok {
		connMutex.Unlock()
		return "invalid connection handle"
	}
	delete(connections, conn)
	connMutex.Unlock()

	err := c.Close()
	if err != nil {
		return err.Error()
	}
	return ""
}

// CloseListener closes the listener.
func CloseListener(listener int) string {
	listenMutex.Lock()
	l, ok := listeners[listener]
	if !ok {
		listenMutex.Unlock()
		return "invalid listener handle"
	}
	delete(listeners, listener)
	listenMutex.Unlock()

	err := l.Close()
	if err != nil {
		return err.Error()
	}
	return ""
}

// SetReadTimeout sets read timeout in milliseconds for a connection.
func SetReadTimeout(conn int, timeoutMs int) string {
	connMutex.Lock()
	c, ok := connections[conn]
	connMutex.Unlock()

	if !ok {
		return "invalid connection handle"
	}

	var deadline time.Time
	if timeoutMs > 0 {
		deadline = time.Now().Add(time.Duration(timeoutMs) * time.Millisecond)
	}

	err := c.SetReadDeadline(deadline)
	if err != nil {
		return err.Error()
	}
	return ""
}

File: runtime/net/rust/net_runtime.rs
use std::collections::HashMap;
use std::net::{TcpListener, TcpStream, ToSocketAddrs};
use std::sync::Mutex;
use std::time::Duration;

// Connection and listener handle management
lazy_static::lazy_static! {
    static ref CONNECTIONS: Mutex<HashMap<i32, TcpStream>> = Mutex::new(HashMap::new());
    static ref LISTENERS: Mutex<HashMap<i32, TcpListener>> = Mutex::new(HashMap::new());
    static ref NEXT_CONN_HANDLE: Mutex<i32> = Mutex::new(1);
    static ref NEXT_LISTEN_HANDLE: Mutex<i32> = Mutex::new(1);
}

// Dial connects to the address on the named network.
pub fn Dial(network: String, address: String) -> (i32, String) {
    if network != "tcp" && network != "tcp4" && network != "tcp6" {
        return (-1, format!("unsupported network: {}", network));
    }

    // Resolve address
    let addrs: Vec<_> = match address.to_socket_addrs() {
        Ok(addrs) => addrs.collect(),
        Err(e) => return (-1, e.to_string()),
    };

    if addrs.is_empty() {
        return (-1, "failed to resolve address".to_string());
    }

    match TcpStream::connect(&addrs[..]) {
        Ok(stream) => {
            let mut next = NEXT_CONN_HANDLE.lock().unwrap();
            let handle = *next;
            *next += 1;

            let mut connections = CONNECTIONS.lock().unwrap();
            connections.insert(handle, stream);
            (handle, String::new())
        }
        Err(e) => (-1, e.to_string()),
    }
}

// Listen announces on the local network address.
pub fn Listen(network: String, address: String) -> (i32, String) {
    if network != "tcp" && network != "tcp4" && network != "tcp6" {
        return (-1, format!("unsupported network: {}", network));
    }

    match TcpListener::bind(&address) {
        Ok(listener) => {
            let mut next = NEXT_LISTEN_HANDLE.lock().unwrap();
            let handle = *next;
            *next += 1;

            let mut listeners = LISTENERS.lock().unwrap();
            listeners.insert(handle, listener);
            (handle, String::new())
        }
        Err(e) => (-1, e.to_string()),
    }
}

// Accept waits for and returns the next connection to the listener.
pub fn Accept(listener: i32) -> (i32, String) {
    let listener_ref = {
        let listeners = LISTENERS.lock().unwrap();
        match listeners.get(&listener) {
            Some(l) => {
                match l.try_clone() {
                    Ok(cloned) => cloned,
                    Err(e) => return (-1, e.to_string()),
                }
            }
            None => return (-1, "invalid listener handle".to_string()),
        }
    };

    match listener_ref.accept() {
        Ok((stream, _addr)) => {
            let mut next = NEXT_CONN_HANDLE.lock().unwrap();
            let handle = *next;
            *next += 1;

            let mut connections = CONNECTIONS.lock().unwrap();
            connections.insert(handle, stream);
            (handle, String::new())
        }
        Err(e) => (-1, e.to_string()),
    }
}

// Read reads up to size bytes from the connection.
pub fn Read(conn: i32, size: i32) -> (Vec<u8>, i32, String) {
    use std::io::Read as IoRead;
    let mut connections = CONNECTIONS.lock().unwrap();
    match connections.get_mut(&conn) {
        Some(stream) => {
            let mut buf = vec![0u8; size as usize];
            match stream.read(&mut buf) {
                Ok(n) => {
                    buf.truncate(n);
                    (buf, n as i32, String::new())
                }
                Err(e) => (Vec::new(), 0, e.to_string()),
            }
        }
        None => (Vec::new(), 0, "invalid connection handle".to_string()),
    }
}

// Write writes data to the connection.
pub fn Write(conn: i32, data: Vec<u8>) -> (i32, String) {
    use std::io::Write as IoWrite;
    let mut connections = CONNECTIONS.lock().unwrap();
    match connections.get_mut(&conn) {
        Some(stream) => match stream.write_all(&data) {
            Ok(_) => (data.len() as i32, String::new()),
            Err(e) => (0, e.to_string()),
        },
        None => (0, "invalid connection handle".to_string()),
    }
}

// Close closes the connection.
pub fn Close(conn: i32) -> String {
    let mut connections = CONNECTIONS.lock().unwrap();
    match connections.remove(&conn) {
        Some(stream) => {
            drop(stream);
            String::new()
        }
        None => "invalid connection handle".to_string(),
    }
}

// CloseListener closes the listener.
pub fn CloseListener(listener: i32) -> String {
    let mut listeners = LISTENERS.lock().unwrap();
    match listeners.remove(&listener) {
        Some(l) => {
            drop(l);
            String::new()
        }
        None => "invalid listener handle".to_string(),
    }
}

// SetReadTimeout sets read timeout in milliseconds for a connection.
pub fn SetReadTimeout(conn: i32, timeout_ms: i32) -> String {
    let mut connections = CONNECTIONS.lock().unwrap();
    match connections.get_mut(&conn) {
        Some(stream) => {
            let timeout = if timeout_ms > 0 {
                Some(Duration::from_millis(timeout_ms as u64))
            } else {
                None
            };

            match stream.set_read_timeout(timeout) {
                Ok(_) => String::new(),
                Err(e) => e.to_string(),
            }
        }
        None => "invalid connection handle".to_string(),
    }
}
